src/airy.jl

```diff
@@ -1,107 +1,523 @@
+#    and downward recurrence to get besseli(1/3, x). Another difficult region is when -10<x<-5 and the imaginary part is close to zero.
+#    In this region we use rotation (see connection formulas http://dlmf.nist.gov/9.2.v) to shift to different region of complex plane
+#    where algorithms show good convergence. If imag(z) == zero then we use the reflection identities to compute in terms of bessel functions.
+#    In general, the cutoff regions compared to [1] are different to provide full double precision accuracy and to prioritize using the power series
+#    and asymptotic expansion compared to other approaches.
 #
-    
+# [1] Jentschura, Ulrich David, and E. Lötstedt. "Numerical calculation of Bessel, Hankel and Airy functions." 
+#     Computer Physics Communications 183.3 (2012): 506-519.
+
 """
     airyai(x)
 Airy function of the first kind ``\\operatorname{Ai}(x)``.
 """
-airyai(x::Real) = _airyai(float(x))
+airyai(z::Number) = _airyai(float(z))
 
 _airyai(x::Float16) = Float16(_airyai(Float32(x)))
+_airyai(z::ComplexF16) = ComplexF16(_airyai(ComplexF32(z)))
+
+function _airyai(z::ComplexOrReal{T}) where T <: Union{Float32, Float64}
+    if ~isfinite(z)
+        isnan(z) && return z
+        if abs(angle(z)) < 2π/3
+            return exp(-z)
+        else
+            return 1 / z
+        end
+    end
+    x, y = real(z), imag(z)
+    zabs = abs(z)
 
-function _airyai(x::T) where T <: Union{Float32, Float64}
-    isnan(x) && return x
-    x_abs = abs(x)
-    z = 2 * x_abs^(T(3)/2) / 3
+    airy_large_argument_cutoff(z) && return airyai_large_argument(z)
+    airyai_power_series_cutoff(x, y) && return airyai_power_series(z)
 
     if x > zero(T)
-        return isinf(z) ? zero(T) : sqrt(x / 3) * besselk(one(T)/3, z) / π
-    elseif x < zero(T)
-        Jv, Yv = besseljy_positive_args(one(T)/3, z)
-        Jmv = (Jv - sqrt(T(3)) * Yv) / 2
-        return isinf(z) ? throw(DomainError(x, "airyai(x) is not defined.")) : sqrt(x_abs) * (Jmv + Jv) / 3
-    elseif iszero(x)
-        return T(0.3550280538878172)
+        # use relation to besselk (http://dlmf.nist.gov/9.6.E1)
+        zz = 2 * z^(T(3)/2) / 3
+        return sqrt(z / 3) * besselk_continued_fraction_shift(one(T)/3, zz) / π
+    else
+        # z is close to the negative real axis
+        # for imag(z) == 0 use reflection to compute in terms of bessel functions of first kind (http://dlmf.nist.gov/9.6.E5)
+        # use computation for real numbers then convert to input type for stability
+        # for imag(z) != 0 use rotation identity (http://dlmf.nist.gov/9.2.E14)
+        if iszero(y)
+            xabs = abs(x)
+            xx = 2 * xabs^(T(3)/2) / 3
+            Jv, Yv = besseljy_positive_args(one(T)/3, xx)
+            Jmv = (Jv - sqrt(T(3)) * Yv) / 2
+            return convert(eltype(z), sqrt(xabs) * (Jmv + Jv) / 3)
+        else
+            return exp(pi*im/3) * _airyai(-z*exp(pi*im/3))  + exp(-pi*im/3) * _airyai(-z*exp(-pi*im/3))
```

Comment:
I don't trust the `exp(pi*im/3)` to constant fold (especially pre 1.8). `cispi(1/3)` is probably a bit better.

Reply:
This is a good catch! Thank you